File: examples/language-samples/sample.ts
// Real TypeScript: generics, unions, decorators, async/await.
interface Repository<T extends { id: number }> {
    find(id: number): Promise<T | undefined>;
}

type Status = "pending" | "active" | "closed";

class UserRepo implements Repository<{ id: number; name: string }> {
    private cache = new Map<number, { id: number; name: string }>();

    async find(id: number) {
        if (this.cache.has(id)) return this.cache.get(id);
        const res = await fetch(`/api/users/${id}`);
        return (await res.json()) as { id: number; name: string };
    }
}

const status: Status = "active";
console.log(`status is ${status}`, new UserRepo());

File: examples/language-samples/sample.swift
// Real Swift: protocols, generics, async/await, property wrappers, enums.
import Foundation

protocol Repository {
    associatedtype Item
    func find(id: Int) async throws -> Item?
}

enum LoadState<T> {
    case idle
    case loading
    case loaded(T)
    case failed(Error)
}

struct Employee: Codable, Identifiable {
    let id: Int
    var name: String
    var salary: Decimal

    var isSenior: Bool { salary > 100_000 }
}

final class EmployeeStore: Repository, @unchecked Sendable {
    private var cache: [Int: Employee] = [:]
    private let lock = NSLock()

    func find(id: Int) async throws -> Employee? {
        lock.lock()
        defer { lock.unlock() }

        if let cached = cache[id] { return cached }

        guard let url = URL(string: "https://example.com/api/\(id)") else {
            return nil
        }
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(Employee.self, from: data)
    }
}
